specs/agents/tracing-instrumentation-azure-functions.md

```diff
@@ -0,0 +1,114 @@
+| `faas.name` | e.g. `MyFunctionApp/MySampleTrigger` | The function app name and the function name, using this format: `<FUNCTION_APP_NAME>/FUNCTION_NAME>`. | *generic context* |
+| `faas.id` | e.g. `/subscriptions/d2ba53be-0815-4...` | The [fully qualified resource ID](https://learn.microsoft.com/en-us/rest/api/resources/resources/get-by-id) of the Azure Function, which has this format: `/subscriptions/<SUBSCRIPTION_GUID>/resourceGroups/<RG>/providers/Microsoft.Web/sites/<FUNCAPP>/functions/<FUNC>` | *generic context*, environment |
+| `faas.trigger.type` | `other` | The trigger type. Use `other` if the trigger type is unknown or cannot be specified. | More concrete triggers are `http`, `pubsub`, `datasource`, `timer` (see specific triggers below). |
+| `faas.execution` | `203621a2-62f...` | The unique invocation id of the function. | *generic context* |
+| `faas.coldstart` | `true`/`false` | A boolean value indicating whether this function invocation was a cold start or not. See the [Deriving cold starts](#deriving-cold-starts) section below. |
+
+### Metadata
+
+Generic cloud metadata fetching SHOULD be disabled when the agent is running in an Azure Functions app. It will not
+yield any useful data and only incur additional cold start overhead due to the non-existent HTTP metadata endpoints.
+
+Agents SHOULD rather:
+
+* Check for the existence of the `FUNCTIONS_WORKER_RUNTIME` environment variable that indicates that the agent is running
+in an Azure Function application.
+* Perform specific cloud metadata detection for Azure Functions (see [Metadata](./metadata.md)).
+
+In addition to those described in [Metadata](./metadata.md), following metadata fields are relevant for Azure Functions:
+
+| Field | Value | Description | Source |
+| - | - | - | - |
+| `service.name` | e.g. `MyFunctionApp` | If the service name is *explicitly* specified through the `service_name` agent config option, use that value. Otherwise, use the name of the Function App. | If `service_name` is not specified, use `WEBSITE_SITE_NAME`. |
+| `service.name` | e.g. `MyFunctionApp` | If the service name is *explicitly* specified through the `service_name` agent config option, use that value. Otherwise, use the name of the Function App. | If `service_name` is not specified, use `WEBSITE_SITE_NAME`. |
+| `service.framework.name` | `Azure Functions` | Constant value for the framework name. | |
+| `service.framework.version` | e.g. `~4` | Version of the Azure Functions runtime. | `FUNCTIONS_EXTENSION_VERSION` |
+| `service.runtime.name`| e.g. `dotnet-isolated` | The language worker runtime (see [here](https://learn.microsoft.com/en-us/azure/azure-functions/functions-app-settings#functions_worker_runtime)). | `FUNCTIONS_WORKER_RUNTIME` |
+| `service.node.configured_name` | e.g. `25d4009bce1d ...` | Unique ID of the VM instance. | `WEBSITE_INSTANCE_ID` ([Azure docs](https://learn.microsoft.com/en-us/azure/app-service/reference-app-settings#scaling)) |
```

Comment:
Hi @trentm,
@gregkalapos [made me aware of](https://github.com/elastic/apm-agent-dotnet/pull/1959#discussion_r1055837164) the fact that this property can be set via [configuration](https://www.elastic.co/guide/en/apm/agent/dotnet/current/config-core.html#config-service-node-name).

So, we could:
* Take take the same approach as with `service.name` and only use the environment variable if no config option is present. However, I don't think the config option makes a lot of sense in an Azure Functions context.
* Use `WEBSITE_INSTANCE_ID` in any case and maybe print a log message if this would override the config option.

My preferred option would be the latter. WDYT?

Reply:
@z1c0  @gregkalapos 

I looked back at the history of `metadata.service.node.configured_name`.
It was added in APM Server 7.5 (released December 2, 2019). Discussion at: https://github.com/elastic/apm/issues/141
In [this comment](https://github.com/elastic/apm/issues/141#issuecomment-530280227) the intake field name `configured_name` was suggested -- to be similar to the already existing `metadata.system.configured_hostname` and `metadata.system.detected_hostname` -- saying:

> This allows us to add a `service.node.[detected]_name` in the future.

I think Azure Functions and AWS Lambda are reasonable cases for a new `metadata.service.node.detected_name`. However, that would be separate work, and would obviously then require a newer APM Server version. So I'll ignore that idea for now.

* * *

I agree that using the `service_node_name` config var doesn't make sense for a serverless system like Azure Functions.  My (somewhat unsure) understanding from the above issue is that the intent of `service_node_name` is for multiple instances of the same service running on the same node and that setting it could result in some harm -- from [this comment](https://github.com/elastic/apm/issues/141#issuecomment-529822151)]:

> I have the fear that making the name configurable could to more harm than good. We heavily rely on the value being unique across a cluster.

Therefore:

- I think `WEBSITE_INSTANCE_ID`, if it exists, should "win" over a configured `service_node_name`.
- Possibly the `service_node_name` config var should explicitly be ignored in an Azure Function or Lambda. (If we think so, this should be a separate spec PR. If not, then at the least, agent documentation for the `service_node_name` config var should say to not use it in serverless configs.)

Thoughts?